persist: migrate structured arrow data as necessary on read

danhhz opened the pull request:
We now have both a write schema and a read schema at read time. The
write schema is specified by an id which indexes into a map in durable
persist state, but is sometimes missing because of backward
compatibility. The read schema is always present but might be missing an
id (e.g. if we use this mechanism for projection pushdown, but also for
other various edge cases).

The vast majority common case is that these schemas will be the same and
no migration is necessary. If we're missing the write schema, we _have_
to use the legacy Codec to decode the data. If we have both and they're
not the same, construct (and cache) migration logic for converting from
the write schema to the read schema.

Note that, counter-intuitively, the write schema could be from "after"
the read schema. This is because of the following scenario:

- A persist source is started on some shard at schema 0.
- The shard is migrated to schema 1. We don't want to restart the
  dataflow because it would be too disruptive.
- Data is written at schema 1 and the persist source encounters it.
- We solve this by only allowing migrations that are both backward and
  forward compatible with all registered migrations.
- In practice, this just means a field id cannot be reused after it is
  dropped. Nothing special to implement here right now because we don't
  (yet) support dropping columns in `compare_and_evolve_schema`.

Touches #16625

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

The first commit is pure refactor, so definitely look at the second commit on its own to reduce noise.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

danhhz commented on src/persist-client/src/schema.rs:
```diff
@@ -76,4 +390,334 @@ mod tests {
+        fn decode<'a>(buf: &'a [u8], schema: &Self::Schema) -> Result<Self, String> {
+            let buf = std::str::from_utf8(buf).map_err(|err| err.to_string())?;
+            let mut ret = buf.split(",").map(|x| x.to_owned()).collect::<Vec<_>>();
+            // Fill in nulls or drop columns to match the requested schema.
```
@ParkMyCar We might want to make the same drop columns fix in the Codec impl for Row/SourceData? Otherwise the scenario where a reader is still looking at an old schema wouldn't work

def- reviewed (commented):
I have triggered a nightly run: https://buildkite.com/materialize/nightly/builds/9319

danhhz commented:
FYI, this is mostly disabled as it won't be tickled until `compare_and_evolve_schema` starts getting called.

danhhz commented:
@bkirwi @ParkMyCar friendly ping :)

bkirwi commented:
Yeah, sorry, I've been having a tough time making a dent in this one! I may ask you for some time tomorrow.

ICYMI, that `subscribe_shard_register_forget` unit test failure seems both reproducible and unique to the branch as far as I can tell.

danhhz commented:
> ICYMI, that subscribe_shard_register_forget unit test failure seems both reproducible and unique to the branch as far as I can tell.
Nah, I saw it. I just sometimes bundle fixes for CI failures in with a round of code review feedback to avoid context switching so much

ParkMyCar commented on src/persist-client/src/schema.rs:
```diff
@@ -76,4 +390,334 @@ mod tests {
+            .unwrap();
+        assert_eq!(write0.write_schemas.id.unwrap(), SchemaId(0));
+
+        // Not backward compatible
+        let res = client
+            .compare_and_evolve_schema::<Strings, (), u64, i64>(
+                shard_id,
+                SchemaId(0),
+                &StringsSchema(vec![]),
+                &UnitSchema,
+                d.clone(),
+            )
+            .await
+            .unwrap();
+        assert_eq!(res, CaESchema::Incompatible);
```
In this case are we just dropping the one column? Which should be compatible in the future but isn't now? Apologies if I'm missing something!

bkirwi commented:
I see Parker's already looked things over, but here's the stuff I was chewing over yesterday:
- We can fail to migrate structured data when we don't have a writer schema for it. However, we always _do_ have a datatype for it - we can just grab it from the array - and migration in practice just requires the datatype. Do we care about discarding / rebuilding these arrays when in theory they could be migrated? (Will two schemas with the same output datatype ever have different migrations?)
  - Removing the `PartMigration::Codec` would simplify the branching a little bit - nice but not important.
  - Using datatypes instead of schemas to cache migrations could also solve some issues with missing reader ids.
- We cache a potentially-quadratic number of migrations and never evict the cache. Does this matter?
  - This risk seems mitigated largely by the fact that ad-hoc reader schemas don't have ids and currently can't be cached... so N is just the number of writer schemas, I think, which should not be huge. (But if N is small, calculating N migrations per read op seems not super expensive.)
  - I think it makes sense to me that we'd cache schemas - they're effectively part of state. But the cost-benefit of caching all _pairs_ of schemas in state seems less clear (vs. at the reader level or something); I don't have a good enough intuition for the usage patterns I think.
- You mention reader schemas can be older that writer schemas. Does this cause a problem for backwards-but-not-forwards-compatible migrations like making a column nullable? (It looks like it might panic but I haven't checked.)

I don't know that any of that's blocking, and I think I'm comfortable going ahead if you've got more clarity on this stuff than me!

danhhz reviewed (commented):
> * We can fail to migrate structured data when we don't have a writer schema for it. However, we always _do_ have a datatype for it - we can just grab it from the array - and migration in practice just requires the datatype. Do we care about discarding / rebuilding these arrays when in theory they could be migrated? (Will two schemas with the same output datatype ever have different migrations?)
>   
>   * Removing the `PartMigration::Codec` would simplify the branching a little bit - nice but not important.
>   * Using datatypes instead of schemas to cache migrations could also solve some issues with missing reader ids.

I think it's a good idea to use the SchemaId as the cache key where we can, that's one of the benefits of having SchemaIds in the first place. I have been vaguely picturing Uuids as the key when we're missing a SchemaId, but DataType is potentially much better, particularly for when the missing SchemaId is the writer one. This PR already feels sufficiently large to me, so inclined to call that part of #29456 if that's okay

> * We cache a potentially-quadratic number of migrations and never evict the cache. Does this matter?
>   
>   * This risk seems mitigated largely by the fact that ad-hoc reader schemas don't have ids and currently can't be cached... so N is just the number of writer schemas, I think, which should not be huge. (But if N is small, calculating N migrations per read op seems not super expensive.)
>   * I think it makes sense to me that we'd cache schemas - they're effectively part of state. But the cost-benefit of caching all _pairs_ of schemas in state seems less clear (vs. at the reader level or something); I don't have a good enough intuition for the usage patterns I think.

I'd been picturing adding an LRU for the n^2 problem if we hit it, but per-handle caches for migrations is much better. Gonna type that up and push it here.

> * You mention reader schemas can be older that writer schemas. Does this cause a problem for backwards-but-not-forwards-compatible migrations like making a column nullable? (It looks like it might panic but I haven't checked.)

Yeah, no idea what to do here: MakeNullable is annoying

danhhz commented:
There's a few core persist bits that snuck in here (like the shard_source_fetch task mentioned below) so gonna kick off a nightlies, just in case. Otherwise, I think I've addressed the review feedback, so this should be good to go!

> I'd been picturing adding an LRU for the n^2 problem if we hit it, but per-handle caches for migrations is much better. Gonna type that up and push it here.

Done! This turned out great.

> ICYMI, that subscribe_shard_register_forget unit test failure seems both reproducible and unique to the branch as far as I can tell.

This turned out to be the addition of more work into `create_batch_fetcher` tickling the old bug where timely async operators don't drive ready futures promptly. So I put the usual fix in shard_source_fetch when creating the fetcher because it seemed like a not-bad idea anyway.

danhhz commented:
Looks like the test_storage_usage_collection_interval timeout [is an existing flake](https://buildkite.com/materialize/test/builds/90147).

TFTRs!
